x-pack/plugins/security_solution/server/endpoint/routes/actions/isolation.ts

```diff
@@ -157,22 +157,18 @@ export const isolationRequestHandler = function (
       });
     }
 
-    const commentLines: string[] = [];
-
-    commentLines.push(`${isolate ? 'I' : 'Uni'}solate action was sent to the following Agents:`);
-    // lines of markdown links, inside a code block
-
-    commentLines.push(`${agentIDs.map((a) => `- [${a}](/app/fleet#/agents/${a})`).join('\n')}`);
-    if (req.body.comment) {
-      commentLines.push(`\n\nWith Comment:\n> ${req.body.comment}`);
-    }
-
+    // TODO: This sets the comment type, we'll need to use our new type.
     caseIDs.forEach(async (caseId) => {
       (await endpointContext.service.getCasesClient(req)).attachments.add({
         caseId,
         comment: {
-          comment: commentLines.join('\n'),
-          type: CommentType.user,
+          type: CommentType.actions,
+          comment: req.body.comment || '',
+          actions: {
+            endpointId: req?.body?.endpoint_ids[0],
```

Comment:
@pzl does this look right?
This API can be used with multiple endpoint IDs (or no endpoint IDs at all), but we are only using the first endpoint ID here?

Also,
If called only with Agent IDs, then we should do the reverse mapping - no? to get an endpoint ID(s) for the agent ID(s).


_Also - FYI: my PR refactors some pieces of this API here to fix a bug where there was a delay between this API returning and the case attachments being created (`await` inside of a class iteration method does not actually await. My changes [here](https://github.com/elastic/kibana/pull/103160/files#diff-7f758056adbf14d1692df37633f8b0720a1495d6bed52810541d7dd65fb90a7aR171) for reference._

Reply:
agree, this doesn't seem right.

The inputs here are `0 - N` endpoint IDs and `0 - M` agent IDs. as long as `N+M > 0`. And then it is all turned into agent IDs. And that resulting list is what we need to work from. Don't use the lists from the input.

As well, grabbing `[0]` ignores the rest of the list.

Taking in a single EndpointID into this comment saved object seems to be a problem. We can have, say, 50 IDs being un/isolated here. Do we need to attach 50 comments for each machine, in each case? Or should we have these comments accept multiple machines (I know this differs from mocks we are working from).

The other problem is that the comment is accepting _Endpoint_ IDs. and we have _Agent_ IDs. On the backend we can _nearly_ guarantee turning an ep id into an agent ID. The reverse doesn't quite hold. (you could call this API with agent IDs that don't have endpoint configured, etc). Can the comment accept a list of agent IDs? What are we using the IDs for on the other end?

Reply:
@pzl @paul-tavares 
This is my fault for not clarifying.  When @parkiino and I were pairing on this, we gave ourselves enough data to get the UI to render properly and test that.  We knew that we would need some work from Dan on the backend to finish out the rest.

Here are what the comments look like that we need to visualize:
![image](https://user-images.githubusercontent.com/56395104/123442619-1003c480-d5a3-11eb-8b2b-7495da87e6e4.png)

> Can the comment accept a list of agent IDs? What are we using the IDs for on the other end?

The Endpoint link is meant to link out to the Endpoint details on our Administration page so we need a set of Endpoint IDs returned in this field.

> Taking in a single EndpointID into this comment saved object seems to be a problem. We can have, say, 50 IDs being un/isolated here. Do we need to attach 50 comments for each machine, in each case?

I agree with you both, we should save an array of Endpoint IDs here and then on the UI we can decide what to do with it.  The common case for the Cases Comment UI is for the user to select an Alert inside the Case and Isolate the single Endpoint associated with it.  We could make the UI smart enough to show something like "User isolated host Win-10 and 5 more" if we receive multiple IDs.  We can expand on this case more later.

The Saved Object defined here: https://github.com/elastic/kibana/pull/102937/files/53e7d3dc83565d1b4c6fe793492d1c98e55e10c9#diff-b299c946efb5112922f55e7612f4e05c57891db35390acfb42735381855cfadcR67

Could be changed to be something like:

```js
export const ActionsCommentRequestRt = rt.type({
  type: rt.literal(CommentType.actions),
  comment: rt.string,
  actions: rt.type({
    hostnames: rt.array(rt.string), // Endpoint hostname field
    endpointIds: rt.array(rt.string, // Endpoint ID so that we can build links to the Administration page
    type: rt.string,
  }),
  owner: rt.string,
});
```

Regarding the inputs, I believe we are always passing in `endpoint_ids` currently since the UI only works from Agents that are running Endpoint (and those Endpoints have triggered an Alert).  If we wanted to make it flexible and accept both types of IDs, then we would need to support a way to get the Endpoint ID + Hostname from `agent_ids`.

2 thoughts:
- In Endpoint metadata docs, we store both `agent.id` (endpoint id) and `elastic.agent.id` (agent id).  Couldn't we perform the lookup there?  We need to fetch Endpoint Hostnames regardless and save it in the Saved Object.
- Also, recently for `7.14.0` Endpoints, the IDs will be the same: https://github.com/elastic/endpoint-dev/pull/9075. Since host isolation will only be supported in `7.14.0`+, maybe the translation logic is no longer required.

We'll work through the details and get this wrapped up next week!